File: Advanced/ch3/uva750.cpp
#include <cstdlib>
#include <cstdio>
#include <cstring>
using namespace std;

int queens[8], numDatasets, a, b, lineCounter;

// TO check previously placed queens
bool place(int row, int col)
{
    for (int prev=0; prev<col; prev++)
    {
        if (queens[prev] == row || 
                (abs(queens[prev] - row) == abs(prev-col)))
            return false;
    }
    return true;
}

void backtrack(int col)
{
    if (col == 8 && queens[b] == a) // candidate solution 
    {
        printf("%2d    %d", ++lineCounter, queens[0]+1);
        for (int j=1; j<8; j++)
            printf("    %d", queens[j]+1);
        printf("\n");
    }
    for (int row=0; row<8; row++)
    {
        // If can place the queen at this row and col,
        // recurse
        if (place(row, col))
        {
            queens[col]=row;
            backtrack(col+1);
        }
    }
}

int main() 
{
    scanf("%d", &numDatasets);
    while (numDatasets--)
    {
        scanf("%d %d", &a, &b);
        a--; b--;
        memset(queens, 0, sizeof queens);
        lineCounter = 0;
        printf("SOLN    COLUMN\n");
        printf("#   1 2 3 4 5 6 7 8\n\n");
        backtrack(0);   // Generate all possible 8! candidate solutions
        if (numDatasets)
            printf("\n");
    }
}

File: 02_Advanced/ch3/uva750.java
// UVA750 - 8 Queens Problem

import java.util.*;

class Main {
    private static int[] row = new int[9];
    private static int TC, a, b, lineCounter;

    private static boolean place(int col, int testrow) {
        for (int prev=1; prev<col; prev++)
            if (row[prev] == testrow || (Math.abs(row[prev]-testrow) == Math.abs(prev-col)))
                return false;
        return true;
    }

    private static void backtrack(int col) {
        for (int testrow=1; testrow <= 8; testrow++) {
            if (place(col, testrow)) {
                row[col] = testrow;
                if (col == 8 && row[b] == a) {
                    // Solution found 
                    System.out.printf("%2d      %d", ++lineCounter, row[1]);
                    for (int j=2; j<=8; j++)
                        System.out.printf(" %d", row[j]);
                    System.out.printf("\n");
                }
                else {
                    backtrack(col+1);
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        TC = sc.nextInt();
        while (TC-- > 0) {
            a = sc.nextInt();
            b = sc.nextInt();
            for (int i=0; i<9; i++)
                row[i] = 0;
            lineCounter = 0;

            System.out.printf("SOLN     COLUMN\n");
            System.out.printf(" #   1 2 3 4 5 6 7 8\n\n");
            backtrack(1);
            if (TC > 0)
                System.out.printf("\n");
        }
    }
}
